File: src/app/match/[id]/gestureHints.ts
/**
 * First-time gesture hints for the player, per device (localStorage).
 *
 * The contract, following the swipe-nudge philosophy and the research
 * behind it: a hint appears at most twice, one hint at a time, and dies
 * forever the first time its gesture is actually used — feedback does
 * the long-term teaching, hints only cover discovery. The gestures
 * sheet (GesturesSheet.tsx) is the replay path, so nothing ever needs
 * to nag.
 */

export type HintName = "dtap" | "hold" | "score" | "notebubble";

const KEY = "ponglens:gesture-hints";
const MAX_SHOWS = 2;

interface HintState {
  shown: Partial<Record<HintName, number>>;
  done: Partial<Record<HintName, boolean>>;
}

function read(): HintState {
  try {
    const raw = window.localStorage.getItem(KEY);
    if (raw) return JSON.parse(raw) as HintState;
  } catch {
    // treat as fresh
  }
  return { shown: {}, done: {} };
}

function write(s: HintState) {
  try {
    window.localStorage.setItem(KEY, JSON.stringify(s));
  } catch {
    // private mode: hints just repeat, which is harmless
  }
}

export function hintEligible(name: HintName): boolean {
  const s = read();
  return !s.done[name] && (s.shown[name] ?? 0) < MAX_SHOWS;
}

export function markHintShown(name: HintName) {
  const s = read();
  s.shown[name] = (s.shown[name] ?? 0) + 1;
  write(s);
}

/** The gesture was used for real: its hint never shows again. */
export function markHintDone(name: HintName) {
  const s = read();
  if (s.done[name]) return;
  s.done[name] = true;
  write(s);
}

File: ios/PongLens/PongLens/Core/GestureHints.swift
import Foundation

/// First-time gesture hints for the player, per device. Port of
/// src/app/match/[id]/gestureHints.ts.
///
/// The contract: a hint appears at most twice, one hint at a time, and dies
/// forever the first time its gesture is actually used. Feedback does the
/// long-term teaching; hints only cover discovery, and the gestures sheet
/// is the replay path — so nothing ever needs to nag.
enum GestureHint: String, CaseIterable {
    /// Double-tap the sides to skip a rally.
    case doubleTap = "dtap"
    /// Hold a side for temporary speed.
    case hold
    /// The first auto-pause in Keep score.
    case score
}

enum GestureHints {
    private static let shownKey = "ponglens.gestureHints.shown"
    private static let doneKey = "ponglens.gestureHints.done"
    private static let maxShows = 2

    private static var shown: [String: Int] {
        get { UserDefaults.standard.dictionary(forKey: shownKey) as? [String: Int] ?? [:] }
        set { UserDefaults.standard.set(newValue, forKey: shownKey) }
    }

    private static var done: [String: Bool] {
        get { UserDefaults.standard.dictionary(forKey: doneKey) as? [String: Bool] ?? [:] }
        set { UserDefaults.standard.set(newValue, forKey: doneKey) }
    }

    static func eligible(_ hint: GestureHint) -> Bool {
        done[hint.rawValue] != true && (shown[hint.rawValue] ?? 0) < maxShows
    }

    static func markShown(_ hint: GestureHint) {
        var s = shown
        s[hint.rawValue] = (s[hint.rawValue] ?? 0) + 1
        shown = s
    }

    /// The gesture was used for real: its hint never shows again.
    static func markDone(_ hint: GestureHint) {
        guard done[hint.rawValue] != true else { return }
        var d = done
        d[hint.rawValue] = true
        done = d
    }

    /// What to teach on this open, if anything: the double tap first, then
    /// hold-for-speed once the double tap is learned or spent.
    static func nextWatchHint() -> GestureHint? {
        if eligible(.doubleTap) { return .doubleTap }
        if eligible(.hold) { return .hold }
        return nil
    }

    static func message(_ hint: GestureHint) -> String {
        switch hint {
        case .doubleTap: "Double-tap the sides to skip a point, the middle to see it again"
        case .hold: "Hold a side for slow motion or double speed"
        case .score: "Tap who won this point"
        }
    }
}
